quantum/debounce/eager_pr.c

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2019 Alex Ong<the.onga@gmail.com>
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*
+Basic per-row algorithm. Uses an 8-bit counter per row.
+After pressing a key, it immediately changes state, and sets a counter.
+No further inputs are accepted until DEBOUNCE milliseconds have occurred.
+*/
+
+#include "matrix.h"
+#include "timer.h"
+#include "quantum.h"
+#include <stdlib.h>
+
+#ifndef DEBOUNCE
+  #define DEBOUNCE 5
+#endif
+
+
+#define debounce_counter_t uint8_t
+
+static debounce_counter_t *debounce_counters;
+
+#define DEBOUNCE_ELAPSED 251
+#define MAX_DEBOUNCE (DEBOUNCE_ELAPSED - 1)
+
+void update_debounce_counters(uint8_t num_rows, uint8_t current_time);
+void transfer_matrix_values(matrix_row_t raw[], matrix_row_t cooked[], uint8_t num_rows, uint8_t current_time);
+
+//we use num_rows rather than MATRIX_ROWS to support split keyboards
+void debounce_init(uint8_t num_rows)
+{
+  debounce_counters = (debounce_counter_t*)malloc(num_rows*sizeof(debounce_counter_t));  
```

Comment:
not a fan of using malloc in an embedded platform. I get the idea behind it, but is it really needed and potentially used by a keyboard right this moment, or a "hey this would be cool" idea that may or may not be used? If it is the later, I would stick with compile time defined array and switched it in the future if a need actually does arise.

Reply:
It's currently unavoidable. There is no unambiguous constant for that value that works for both split and non-split matrix code, so it can't be statically allocated.

Removing it means fixing MATRIX_ROWS for every split keyboard, which is on my list of things to do.

Reply:
What about something along the lines of:

```
#ifndef SPLIT_KEYBOARD
static debounce_counter_t debounce_counters[MATRIX_ROWS];
#else
static debounce_counter_t debounce_counters[ROWS_PER_HAND];
#endif
```

Reply:
hmm, ya I see ROWS_PER_HAND isn't always consistently defined either. Hmm, that's annoying....

Reply:
It's not a good idea to jump through hoops just to eliminate the malloc, especially since it's essentially a static allocation anyway (similar lifetime, allocated at startup, never freed.)

Reply:
ya, but removing malloc does mean you use LDI instead of LDS instruction for accessing data in that array saving 1-2 cpu cycles per loop. Given the talk of bottelnecks above, I thought it was worth bringing up.

Reply:
I think for now we leave the malloc but in the future™ num_rows is better defined so we can stop passing it around everywhere.